# Exercise 1: Game Fun: Customer Acquisition through Digital Advertising

In [191]:
import pandas as pd
import numpy as np
df = pd.read_excel("C:/Users/prath/Documents/Personal/Spring Quarter/BAX 423- Big Data Analytics/Homework 1/GameFun.xlsx")
df.head()

id  test  purchase   site  impressions  income  gender  gamer
0   1956     0         0  site1            0     100       1      0
1  45821     1         0  site1           20      70       1      0
2  59690     1         0  site1           22     100       1      0
3  18851     0         0  site1           13      90       1      0
4  60647     1         0  site1           12      60       1      0

In [192]:
df.shape

(40048, 8)

In [193]:
test = df[df['test']==1]
test.head()

id  test  purchase   site  impressions  income  gender  gamer
1  45821     1         0  site1           20      70       1      0
2  59690     1         0  site1           22     100       1      0
4  60647     1         0  site1           12      60       1      0
5  30908     1         0  site1            0      40       1      0
7   4187     1         0  site1           16      80       1      0

In [194]:
control = df[df['test']==0]
control.head()

id  test  purchase   site  impressions  income  gender  gamer
0    1956     0         0  site1            0     100       1      0
3   18851     0         0  site1           13      90       1      0
6   13352     0         0  site1            0      60       1      0
9   13373     0         0  site1           20      40       1      0
16  13400     0         0  site1            2      50       1      0

### 1.Before evaluating the effect of an experiment, it is important to make sure that the experiment was executed correctly. Check whether the test and control groups are probabilistically equivalent on their observables

#### a.More specific, compare the averages of the income, gender and gamer variables in the test and control groups. You should also report the % difference in the averages. Compute its statistical significance. 

In [234]:
#Income
test_income = test.income.mean()
control_income = control.income.mean()

#Test male and female
total_count_test = test['test'].count()

female_test = test[test['gender']==0]
female_test = female_test['test'].count()
female_test_percent = female_test/total_count_test*100
male_test = test[test['gender']==1]
male_test = male_test['test'].count()
male_test_percent = male_test/total_count_test*100

#Control male and female
total_count_control = control['test'].count()

female_control = control[control['gender']==0]
female_control = female_control['test'].count()
female_control_percent = female_control/total_count_control*100
male_control = control[control['gender']==1]
male_control = male_control['test'].count()
male_control_percent = male_control/total_count_control*100

#Test gamer and non gamer
test_gamer = test[test['gamer']==1]
test_gamer = test_gamer['test'].count()
gamer_test_percent = test_gamer/total_count_test*100
test_nogamer = test[test['gamer']==0]
test_nogamer = test_nogamer['test'].count()
nogamer_test_percent = test_nogamer/total_count_test*100

#Control gamer and non gamer
control_gamer = control[control['gamer']==1]
control_gamer = control_gamer['test'].count()
gamer_control_percent = control_gamer/total_count_control*100


#Difference between the test and control groups for all the above categories
percent_diff_income = (test_income-control_income)*100/test_income
female_diff = female_test_percent-female_control_percent
male_diff = male_test_percent-male_control_percent
gamer_diff = gamer_test_percent-gamer_control_percent

print('% difference of incomes in test and control groups are : ',round(percent_diff_income,2))
print('Difference of female in test and control groups are : ',round(female_diff,2))
print('Difference of male in test and control groups are : ',round(male_diff,2))
print('Difference of gamers in test and control groups are : ',round(gamer_diff,2))

% difference of incomes in test and control groups are :  -0.41
Difference of female in test and control groups are :  0.06
Difference of male in test and control groups are :  -0.06
Difference of gamers in test and control groups are :  -0.05


In [231]:
from scipy import stats
t_value, p_value = stats.ttest_ind(test.income,
                                   control.income)

print('p-value for two tailed test for income in test and control is %f' % p_value)

p-value for two tailed test for income in test and control is 0.128358


In [232]:
t_value, p_value = stats.ttest_ind(test.gender,
                                   control.gender)

print('p-value for two tailed test for gender in test and control is %f' % p_value)

p-value for two tailed test for gender in test and control is 0.906033


In [233]:
t_value, p_value = stats.ttest_ind(test.gamer,
                                   control.gamer)

print('p-value for two tailed test for gamer in test and control is %f' % p_value)

p-value for two tailed test for gamer in test and control is 0.926704


#### b. Briefly comment on what these metrics tell you about probabilistic equivalence for this experiment. 

The test and control groups are designed well with probabilistic equivalence. The average mean and demographics proportions between test and control are statistically equal.

#### c. If you had run this type of analysis BEFORE executing an experiment and found a large difference between test and control groups, what you should do?

If there is a large difference between the test & control groups, it may indicate that we could get better test set performance with more regularization/introducing more bias to the model. But that does not mean that a smaller gap means a better model; it's just that if we have a small or no gap between control and test set performance, we know we are definitely not overfitting so adding regularization/introducing more bias to the model will not help.
However, in order to conduct a better experiment, we would have to re-assign the test and control groups in such a way that they are equivalent probabilistically, demographically and taking all other factors into consideration.

### 2. Evaluate the average purchase rates in the test and control for the following groups. For each comparison, report the average purchase rate for the test, average purchase rate for the control and the absolute difference (not the % difference) between the test and control

### a. Comparison 1: All customers

In [199]:
test_purchase_1 = test[test['purchase']==1].count()
control_purchase_1 = control[control['purchase']==1].count()
test_total_purchases = test.purchase.count()
control_total_purchases = control.purchase.count()
test_avg_purchase_rate_allcust = test_purchase_1/test_total_purchases
control_avg_purchase_rate_allcust = control_purchase_1/control_total_purchases

print('Average purchase rate for all customers in test group (in %): ',round(test_avg_purchase_rate_allcust['purchase']*100,2),'%')
print('Average purchase rate for all customers in control group (in %): ',round(control_avg_purchase_rate_allcust['purchase']*100,2),'%')
abs_diff = test_avg_purchase_rate_allcust['purchase']-control_avg_purchase_rate_allcust['purchase']
print('Absolute difference between all customers in control and test group : ',round(abs_diff,3))

Average purchase rate for all customers in test group (in %):  7.68 %
Average purchase rate for all customers in control group (in %):  3.62 %
Absolute difference between all customers in control and test group :  0.041


### b. Comparison 2: Male vs Female customers

In [200]:
male_test =  test[test['gender']==1]
female_test = test[test['gender']==0]
male_control = control[control['gender']==1]
female_control = control[control['gender']==0]

maletest_purchase_1 = male_test[male_test['purchase']==1].count()
malecontrol_purchase_1 = male_control[male_control['purchase']==1].count()

femaletest_purchase_1 = female_test[female_test['purchase']==1].count()
femalecontrol_purchase_1 = female_control[female_control['purchase']==1].count()

maletest_total_purchases = male_test.purchase.count()
femaletest_total_purchases = female_test.purchase.count()
malecontrol_total_purchases = male_control.purchase.count()
femalecontrol_total_purchases = female_control.purchase.count()

maletest_avg_purchase_rate = maletest_purchase_1/maletest_total_purchases
femaletest_avg_purchase_rate = femaletest_purchase_1/femaletest_total_purchases
malecontrol_avg_purchase_rate = malecontrol_purchase_1/malecontrol_total_purchases
femalecontrol_avg_purchase_rate = femalecontrol_purchase_1/malecontrol_total_purchases

print('Average purchase rate for male customers in test group : ',round(maletest_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for female customers in test group : ',round(femaletest_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for male customers in control group : ',round(malecontrol_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for female customers in control group : ',round(femalecontrol_avg_purchase_rate['purchase']*100,2),'%')

abs_diff_male = maletest_avg_purchase_rate['purchase']-malecontrol_avg_purchase_rate['purchase']
print('Absolute difference between male customers in control and test group : ',round(abs_diff_male,3))

abs_diff_female = femaletest_avg_purchase_rate['purchase']-femalecontrol_avg_purchase_rate['purchase']
print('Absolute difference between female customers in control and test group : ',round(abs_diff_female,3))

Average purchase rate for male customers in test group :  7.46 %
Average purchase rate for female customers in test group :  8.09 %
Average purchase rate for male customers in control group :  3.72 %
Average purchase rate for female customers in control group :  1.87 %
Absolute difference between male customers in control and test group :  0.037
Absolute difference between female customers in control and test group :  0.062


### c. Comparison 3: Gamers vs Non-Gamers Customers

In [201]:
gamer_test =  test[test['gamer']==1]
nongamer_test = test[test['gamer']==0]
gamer_control = control[control['gamer']==1]
nongamer_control = control[control['gamer']==0]

gamertest_purchase_1 = gamer_test[gamer_test['purchase']==1].count()
gamercontrol_purchase_1 = gamer_control[gamer_control['purchase']==1].count()

nongamertest_purchase_1 = nongamer_test[nongamer_test['purchase']==1].count()
nongamercontrol_purchase_1 = nongamer_control[nongamer_control['purchase']==1].count()

gamertest_total_purchases = gamer_test.purchase.count()
nongamertest_total_purchases = nongamer_test.purchase.count()
gamercontrol_total_purchases = gamer_control.purchase.count()
nongamercontrol_total_purchases = gamer_control.purchase.count()

gamertest_avg_purchase_rate = gamertest_purchase_1/gamertest_total_purchases
nongamertest_avg_purchase_rate = nongamertest_purchase_1/nongamertest_total_purchases
gamercontrol_avg_purchase_rate = gamercontrol_purchase_1/gamercontrol_total_purchases
nongamercontrol_avg_purchase_rate = nongamercontrol_purchase_1/nongamercontrol_total_purchases

print('Average purchase rate for gamer customers in test group : ',round(gamertest_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for non-gamer customers in test group : ',round(nongamertest_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for gamer customers in control group : ',round(gamercontrol_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for non-gamer customers in control group : ',round(nongamercontrol_avg_purchase_rate['purchase']*100,2),'%')

abs_diff_gamer = gamertest_avg_purchase_rate['purchase']-gamercontrol_avg_purchase_rate['purchase']
print('Absolute difference between gamer customers in control and test group : ',round(abs_diff_gamer,3))

abs_diff_nongamer = nongamertest_avg_purchase_rate['purchase']-nongamercontrol_avg_purchase_rate['purchase']
print('Absolute difference between non-gamer customers in control and test group : ',round(abs_diff_nongamer,3))

Average purchase rate for gamer customers in test group :  10.45 %
Average purchase rate for non-gamer customers in test group :  3.51 %
Average purchase rate for gamer customers in control group :  3.54 %
Average purchase rate for non-gamer customers in control group :  2.47 %
Absolute difference between gamer customers in control and test group :  0.069
Absolute difference between non-gamer customers in control and test group :  0.01


### d. Comparison 4: Female Gamers vs Male Gamers

In [202]:
malegamer_test =  test[(test['gamer']==1) & (test['gender']==1)]
malegamer_test

id  test  purchase   site  impressions  income  gender  gamer
8      40300     1         1  site1            9      80       1      1
10     13376     1         0  site1           68      60       1      1
13     66400     1         0  site1            1      50       1      1
15     59280     1         0  site1            9      60       1      1
22     10321     1         1  site1          343      50       1      1
...      ...   ...       ...    ...          ...     ...     ...    ...
40035  28922     1         0  site2            2      50       1      1
40038  13324     1         0  site2            1      40       1      1
40040  13310     1         0  site2            8      40       1      1
40042  64401     1         0  site2          185      60       1      1
40046  42279     1         0  site2           19      50       1      1

[10897 rows x 8 columns]

In [203]:
malegamer_test =  test[(test['gamer']==1) & (test['gender']==1)]
femalegamer_test =  test[(test['gamer']==1) & (test['gender']==0)]

malegamer_control = control[(control['gamer']==1) & (control['gender']==1)]
femalegamer_control = control[(control['gamer']==1) & (control['gender']==0)]

malegamertest_purchase_1 = malegamer_test[malegamer_test['purchase']==1].count()
malegamercontrol_purchase_1 = malegamer_control[malegamer_control['purchase']==1].count()

femalegamertest_purchase_1 = femalegamer_test[femalegamer_test['purchase']==1].count()
femalegamercontrol_purchase_1 = femalegamer_control[femalegamer_control['purchase']==1].count()

malegamertest_total_purchases = malegamer_test.purchase.count()
femalegamertest_total_purchases = femalegamer_test.purchase.count()
malegamercontrol_total_purchases = malegamer_control.purchase.count()
femalegamercontrol_total_purchases = femalegamer_control.purchase.count()

malegamertest_avg_purchase_rate = malegamertest_purchase_1/malegamertest_total_purchases
femalegamertest_avg_purchase_rate = femalegamertest_purchase_1/femalegamertest_total_purchases
malegamercontrol_avg_purchase_rate = malegamercontrol_purchase_1/malegamercontrol_total_purchases
femalegamercontrol_avg_purchase_rate = femalegamercontrol_purchase_1/femalegamercontrol_total_purchases

print('Average purchase rate for male-gamer customers in test group (in %): ',round(malegamertest_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for female-gamer customers in test group (in %) : ',round(femalegamertest_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for male-gamer customers in control group (in %): ',round(malegamercontrol_avg_purchase_rate['purchase']*100,2),'%')
print('Average purchase rate for female-gamer customers in control group (in %): ',round(femalegamercontrol_avg_purchase_rate['purchase']*100,2),'%')

abs_diff_malegamer = malegamertest_avg_purchase_rate['purchase']-malegamercontrol_avg_purchase_rate['purchase']
print('Absolute difference between male-gamer customers in control and test group: ',round(abs_diff_malegamer,3))

abs_diff_femalegamer = femalegamertest_avg_purchase_rate['purchase']-femalegamercontrol_avg_purchase_rate['purchase']
print('Absolute difference between female-gamer customers in control and test group: ',round(abs_diff_femalegamer,3))

Average purchase rate for male-gamer customers in test group (in %):  10.14 %
Average purchase rate for female-gamer customers in test group (in %) :  11.01 %
Average purchase rate for male-gamer customers in control group (in %):  3.73 %
Average purchase rate for female-gamer customers in control group (in %):  3.2 %
Absolute difference between male-gamer customers in control and test group:  0.064
Absolute difference between female-gamer customers in control and test group:  0.078


### 3.Assess the expected revenue in the test vs. control for the following comparisons:
### a. Comparison 1: All customers 


In [204]:
test_purchase_rev = test[test['purchase']==1].count()
control_purchase_rev = control[control['purchase']==1].count()
test_total_purchases = test.purchase.count()
control_total_purchases = control.purchase.count()
test_avg_purchase_rate_allcust = test_purchase_rev/test_total_purchases
control_avg_purchase_rate_allcust = control_purchase_rev/control_total_purchases

test_expected_rev = test_purchase_rev*test_avg_purchase_rate_allcust*12.5
control_expected_rev = control_purchase_rev*control_avg_purchase_rate_allcust*37.5

print('The expected revenue in the test group is: ',test_expected_rev['test'])
print('The expected revenue in the control group is: ',control_expected_rev['test'])

The expected revenue in the test group is:  2072.2669182300383
The expected revenue in the control group is:  588.0101614117253


### b. Comparison 4: Female Gamers vs Male Gamers

In [205]:
malegamer_test =  test[(test['gamer']==1) & (test['gender']==1)]
femalegamer_test =  test[(test['gamer']==1) & (test['gender']==0)]

malegamer_control = control[(control['gamer']==1) & (control['gender']==1)]
femalegamer_control = control[(control['gamer']==1) & (control['gender']==0)]

malegamertest_purchase_1 = malegamer_test[malegamer_test['purchase']==1].count()
malegamercontrol_purchase_1 = malegamer_control[malegamer_control['purchase']==1].count()

femalegamertest_purchase_1 = femalegamer_test[femalegamer_test['purchase']==1].count()
femalegamercontrol_purchase_1 = femalegamer_control[femalegamer_control['purchase']==1].count()

malegamertest_total_purchases = malegamer_test.purchase.count()
femalegamertest_total_purchases = femalegamer_test.purchase.count()
malegamercontrol_total_purchases = malegamer_control.purchase.count()
femalegamercontrol_total_purchases = femalegamer_control.purchase.count()

malegamertest_avg_purchase_rate = malegamertest_purchase_1/malegamertest_total_purchases
femalegamertest_avg_purchase_rate = femalegamertest_purchase_1/femalegamertest_total_purchases
malegamercontrol_avg_purchase_rate = malegamercontrol_purchase_1/malegamercontrol_total_purchases
femalegamercontrol_avg_purchase_rate = femalegamercontrol_purchase_1/femalegamercontrol_total_purchases

malegamertest_expected_rev = malegamertest_purchase_1*malegamertest_avg_purchase_rate*12.5
malegamercontrol_expected_rev = malegamercontrol_purchase_1*malegamercontrol_avg_purchase_rate*37.5
femalegamertest_expected_rev = femalegamertest_purchase_1*femalegamertest_avg_purchase_rate*12.5
femalegamercontrol_expected_rev = femalegamercontrol_purchase_1*femalegamercontrol_avg_purchase_rate*37.5

print('The expected revenue by male gamers from test group is: ',round(malegamertest_expected_rev['test'],2))
print('The expected revenue by male gamers from control group is: ',round(malegamercontrol_expected_rev['test'],2))
print('The expected revenue by female gamers from test group is: ',round(femalegamertest_expected_rev['test'],2))
print('The expected revenue by female gamers from control group is: ',round(femalegamercontrol_expected_rev['test'],2))

The expected revenue by male gamers from test group is:  1400.64
The expected revenue by male gamers from control group is:  243.22
The expected revenue by female gamers from test group is:  908.26
The expected revenue by female gamers from control group is:  97.32


### 4.Based on your previous answers, provide a brief recommendation to your management team summarizing the expected financial outcome for Game-Fun.
### a. Should Game-Fun run this promotion again in the future? If no, explain why. If yes, should Game-Fun offer it to all customers or a targeted segment.

The promotion should definitely be run again as we see in the above question that the total expected revenue by the test group is 2012 dollars whereas the revenue from the contorl group is 588 dollars. This clearly shows the promotion is impacting the customers choices.

When we dive deep into the contribution of the test group we see that both the genders who are gamers are being hugely affected by the promotion. 

Again, in the above question we see that in case of male gamers the expected revenue is 243 dollars when 'NO' campaign is run whereas it shoots up to 1400 dollars when the promotion is run. This is almost 3 times increase in the revenue.

Similarly,  in case of female gamers the expected revenue is 97 dollars when 'NO' campaign is run whereas it shoots up to 908 dollars when the promotion is run. This is almost 9 times increase in the revenue.

We see that the increase is the same in both cases and is pretty high. Therefore, in the next promotion focusing on the male and female gamers would ideally tend to bring about more revenue.

# Exercise 2: Non-Compliance in Randomized Experiments

### 1. The first data scientist advised that one should compare the survival rate of babieswhose mothers were offered Vitamin A shots to the survival rate of babies whosemothers were not offered a Vitamin A shot.
#### a. What percent of babies whose mothers were offered Vitamin A shots for their babies died?

In [206]:
df1 =  pd.read_csv("C:/Users/prath/Documents/Personal/Spring Quarter/BAX 423- Big Data Analytics/Homework 1/sommer_deger.csv")
df1.head()

instrument  treatment  outcome
0           0          0        0
1           0          0        0
2           0          0        0
3           0          0        0
4           0          0        0

In [207]:
df1.shape

(23682, 3)

In [208]:
instrument = df1['instrument']
treatment = df1['treatment']
outcome = df1['outcome']

In [209]:
(df1['instrument'] == 1).value_counts()

True     12094
False    11588
Name: instrument, dtype: int64

In [210]:
mother_dose = df1[df1['instrument'] == 1]
mother_nodose = df1[df1['instrument'] == 0]

babies_dose = df1[df1['treatment'] == 1]
babies_nodose = df1[df1['treatment'] == 0]

babies_survived = df1[df1['outcome'] == 0]
babies_died = df1[df1['outcome'] == 1]

In [211]:
mothers_dose = df1[(df1['instrument']==1)]
babies_died = mothers_dose[(mothers_dose['outcome']==1)]
percent_death_mothers_dose = (babies_died['outcome'].count()/mothers_dose['instrument'].count())*100

print("Percent of babies whose mothers were offered Vitamin A shots for their babies died: ", round(percent_death_mothers_dose, 2), "%")

Percent of babies whose mothers were offered Vitamin A shots for their babies died:  0.38 %


#### b. What percent of babies whose mothers were not offered Vitamin A shots for their babies died?

In [212]:
mothers_nodose = df1[(df1['instrument']==0)]
babies_died = mothers_nodose[(mothers_nodose['outcome']==1)]
percent_death_mothers_nodose = (babies_died['outcome'].count()/mothers_nodose['instrument'].count())*100

print("Percent of babies whose mothers were NOT offered Vitamin A shots for their babies died: ", round(percent_death_mothers_nodose, 2), "%")

Percent of babies whose mothers were NOT offered Vitamin A shots for their babies died:  0.64 %


#### c. What is the difference in mortality, and under what assumptions is the difference between these two percentages a valid estimate of the causal impact of receiving vitamin A shots on survival?

The mortality difference is around a quarter of a percent. We've found that if women aren't given the proper dosage, the number of babies who die rises by 25%. However, we can see that some mothers refuse to give their newborns vaccinations despite being offered them. As a result, there is a self-selection bias that is not taken into account in this scenario.

Certain assumptions that could help in this scenario are:
- There is a probabilistic equivalence between the test and control
- No spillovers
- No self-selection bias (as mentioned above)
- All the mothers who are offered vitamin A has actually given the vitamin to the babies

### 2. The second data scientist advised that one should compare the survival rates of babies who received Vitamin A shots to babies who did not receive Vitamin A shots.
#### a. What percent of babies who received Vitamin A shots died?

In [213]:
babies_dose = df1[(df1['treatment']==1)]
babies_died = babies_dose[(babies_dose['outcome']==1)]
percent_death_babies_dose = babies_died['outcome'].count()/babies_dose['treatment'].count()*100

print("Percent of babies who received Vitamin A shots died: ", round(percent_death_babies_dose, 3), "%")

Percent of babies who received Vitamin A shots died:  0.124 %


#### b. What percent of babies who did not receive Vitamin A shots died?

In [214]:
babies_nodose = df1[(df1['treatment']==0)]
babies_died = babies_nodose[(babies_nodose['outcome']==1)]
percent_death_babies_nodose = babies_died['outcome'].count()/babies_nodose['treatment'].count()*100

print("Percent of babies who did NOT receive Vitamin A shots died: ", round(percent_death_babies_nodose, 2), "%")

Percent of babies who did NOT receive Vitamin A shots died:  0.77 %


#### c. What is the difference in mortality, and under what assumptions is the difference between these two percentages a valid estimate of the causal impact of receiving vitamin A shots on survival?

The difference between newborns who received the dose and those who did not is approximately 65%. When given to babies directly, the dose has a greater impact than when given to moms. But there's one more factor to consider: those who didn't get shots are a mix of babies whose mothers were offered shots but didn't take them, as well as babies whose moms were not offered shots and couldn't take them.

When the babies who did not receive injections were those whose moms were not offered shots, the difference between these two percentages can be a reasonable indication of the causal influence of receiving vitamin A shots on survival.

Certain assumptions that could help in this scenario are:
- There is a probabilistic equivalence between the test and control
- No spillovers
- No self-selection bias (as mentioned above)
- All the mothers who are offered vitamin A has actually given the vitamin to the babies

### 3. The third data scientist advised that one should consider only babies whose motherswere offered Vitamin A shots, and compare babies who received shots to babies whodid not receive shots.
#### a. What percent of babies who received Vitamin A shots died?

In [215]:
babies_dose = df1[(df1['treatment']==1)]
babies_died = babies_dose[(babies_dose['outcome']==1)]
percent_death_died_dose = babies_died['outcome'].count()/babies_dose['treatment'].count()*100

print("Percent of babies who received Vitamin A shots died: ", round(percent_death_died_dose, 3), "%")

Percent of babies who received Vitamin A shots died:  0.124 %


#### b. What percent of babies whose mothers were offered Vitamin A shots, but themothers did not accept them, died?

In [216]:
mothers_accepted_dose = mothers_dose[mothers_dose['treatment'] == 1]
mothers_rejected_dose = mothers_dose[mothers_dose['treatment'] == 0]

total_mothers_rejected_dose = mothers_rejected_dose['treatment'].count()
didnt_survive_mothers_rejected_dose = mothers_rejected_dose[mothers_rejected_dose['outcome'] == 1]['treatment'].count()
percent_mothers_rejected_dose_died = didnt_survive_mothers_rejected_dose/total_mothers_rejected_dose*100

print("Percent of babies whose mothers were offered Vitamin A shots, but the mothers did not accept them, died: ", round(percent_mothers_rejected_dose_died, 3), "%")


Percent of babies whose mothers were offered Vitamin A shots, but the mothers did not accept them, died:  1.406 %


#### c. What is the difference in mortality, and under what assumptions is the difference between these two percentages a valid estimate of the causal impact of receiving vitamin A shots on survival?

The difference between these two percentages can be a valid estimate of the causal impact of receiving vitamin A shots on survival if the proportion of the babies (whose mothers were offered shots) and received the shot and the proportion of the babies (whose mothers were offered shots) and did not receive the shot are probabilistically equivalent. But we can not control that in this case.

Certain assumptions that could help in this scenario are:
- There is a probabilistic equivalence between the test and control
- No spillovers
- No self-selection bias (as mentioned above)

### 4. The fourth data scientist suggested the following Wald estimator for the effect of Vitamin A shots on mortality
#### a. Compute the above Wald estimate for the given dataset.

In [220]:
total_mothers_accepted_dose = mothers_accepted_dose['treatment'].count()
total = mother_dose['instrument'].count()
percent_mother_offered_baby_received = total_mothers_accepted_dose*100/total
wald_estimate = (percent_death_mothers_dose - percent_death_mothers_nodose)/percent_mother_offered_baby_received

print("Wald Estimate: ", round(wald_estimate, 4))

Wald Estimate:  -0.0032


#### b. Under what assumptions is this estimate a valid estimate of the causal impact of vitamin A shots on survival?

We adjust for the underestimation of the actual treatment impact by utilizing Wald's estimate. Under the premise that we are underestimating the treatment effect in the first place, the estimation of the causal effect of obtaining vitamin A shots will be valid. It's possible that mothers' self-selection bias for not getting the shot stems from the fact that they were already sick and on the verge of dying. This could lead to an overestimation of the treatment's effectiveness. This is not the case with the control group, which is made up entirely of compliers and thus is unbiased, and randomization would handle such situations. We also assume that there is no self-selection in the treatment group.

Certain assumptions that could help in this scenario are:
- There is a probabilistic equivalence between the test and control
- No spillovers
- No self-selection bias (as mentioned above)
- All the mothers who are offered vitamin A has actually given the vitamin to the babies

#### c. What is the standard error for the intent-to-treat estimate recommended by the first data scientist? What is the standard error for the Wald estimate recommended by the fourth data scientist?

In [227]:
#Regressing Treatment variable on Instrument Variable
import statsmodels.api as sm

y = treatment
x = sm.add_constant(instrument)

model1 = sm.OLS(y,x)

In [228]:
results1 = model1.fit()
print(results1.summary())

                            OLS Regression Results                            
Dep. Variable:              treatment   R-squared:                       0.662
Model:                            OLS   Adj. R-squared:                  0.662
Method:                 Least Squares   F-statistic:                 4.634e+04
Date:                Fri, 08 Apr 2022   Prob (F-statistic):               0.00
Time:                        08:26:05   Log-Likelihood:                -3947.2
No. Observations:               23682   AIC:                             7898.
Df Residuals:                   23680   BIC:                             7915.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -3.76e-15      0.003  -1.42e-12      1.0

In [229]:
y_hat = results1.predict(x)

In [230]:
#Regressing Output on Y_hat
y2 = outcome

x2 = sm.add_constant(y_hat)
model2 = sm.OLS(y2,x2)

results2 = model2.fit()
print(results2.summary())

                            OLS Regression Results                            
Dep. Variable:                outcome   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     7.830
Date:                Fri, 08 Apr 2022   Prob (F-statistic):            0.00514
Time:                        08:26:54   Log-Likelihood:                 29040.
No. Observations:               23682   AIC:                        -5.808e+04
Df Residuals:                   23680   BIC:                        -5.806e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0064      0.001      9.683      0.0

#### i. Which one is larger and why? 

The standard error of the intent-to-treat estimate is larger than the standard error of the wald estimate due to intrinsic bias (self selection bias) in the treatment variable. We attempt to correct the bias in the initial treatment variable in order to forecast the mortality rate from the anticipated intent-to-treat; as a result, we are able to derive a tighter bound (lower standard error) for the Wald's estimate.

As a result, the intent-to-treat estimate's standard error is higher.

#### ii. Why might these standard errors be biased? What information would you ideally want to have to address this bias?

Because of the bias in the data, these standard errors may be skewed. We simply assumed that the data is normally distributed and meets the OLS assumptions, but this is not always the case. As a result, implicit bias in the data, particularly in the treatment variable (where we have non-compliers), biases the sample, which biases our standard errors.

We'll need more information about the non-compliers in the treatment variable to estimate the bias. We'll need to know why the mothers who were offered vaccinations for their children declined them. We will then be able to address the problem by quantifying the bias.